# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [2]:
import pandas as pd

In [3]:
original_data = pd.read_csv('e_commerce.csv')

In [4]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

### 评估数据整齐程度

在这一部分，我将对在上一部分建立的 original_data 这个DataFrame所包含的数据进行评估。
评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

In [5]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
295747    562840     23199                    JUMBO BAG APPLES         4   
326400    565541     21991     BOHEMIAN COLLAGE STATIONERY SET         2   
192635    553462     21156           RETROSPOT CHILDRENS APRON         8   
478987    577126     22553              PLASTERS IN TIN SKULLS        12   
437150    574269     22086     PAPER CHAIN KIT 50'S CHRISTMAS         12   
322810    565234     23198      PANTRY MAGNETIC  SHOPPING LIST         2   
205854    554839     21936            RED RETROSPOT PICNIC BAG         1   
540679    581492     21892  TRADITIONAL WOODEN CATCH CUP GAME          1   
255738    559364     21929      JUMBO BAG PINK VINTAGE PAISLEY        10   
280492    561463     22367     CHILDRENS APRON SPACEBOY DESIGN         3   

             InvoiceDate  UnitPrice  CustomerID         Country  
295747    8/9/2011 16:27       2.08     14527.0  United Kingdom  
326400    9/5/2011 12:00       1.25     14159.0  United Kingdom  
192635   5/17/2011 10:57       1.95     17147.0  United Kingdom  
478987  11/17/2011 19:22       1.65     12406.0         Denmark  
437150   11/3/2011 14:27       2.95     15570.0  United Kingdom  
322810     9/2/2011 9:38       3.29         NaN  United Kingdom  
205854   5/26/2011 17:15       2.95     14056.0  United Kingdom  
540679   12/9/2011 10:03       2.46         NaN  United Kingdom  
255738     7/8/2011 8:48       2.08     17735.0  United Kingdom  
280492   7/27/2011 12:38       1.95     14527.0  United Kingdom

从抽样的10行数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是一次交易，每列是交易相关的各个变量，不存在结构性问题。

### 评估数据干净程度

In [6]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，`Description`、`CustomerID`变量存在缺失值。

`InvoiceDate`数据类型应该是时间，`CustomerID`应该是字符串，应当进行数据格式转换。

### 评估缺失数据

In [9]:
# 根据条件提取出缺失description的观察值。
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

从输出结果来看，缺失的`Description`数据`UnitPrice`都为零，为验证猜想，看是否存在`description`变量确实且`UnitPrice`不为0的数据。

In [13]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

输出结果0条，说明缺失产品数据单价都为0。这些观察值都没有意义，后续可删除。

In [15]:
# 根据条件提取出缺失CustomerID的观察值。
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

客户编号缺失不影响对畅销产品的分析，可以保留。

### 评估不一致数据

是否存在不同值指代同一国家的情况？

In [18]:
original_data["Country"].value_counts().sort_index()

Country
Australia                 1259
Austria                    401
Bahrain                     19
Belgium                   2069
Brazil                      32
Canada                     151
Channel Islands            758
China                      288
Cyprus                     622
Czech Republic              30
Denmark                    389
EIRE                      8196
European Community          61
Finland                    695
France                    8557
Germany                   9495
Greece                     146
Iceland                    182
Israel                     297
Italy                      803
Japan                      358
Lebanon                     45
Lithuania                   35
Malta                      127
Netherlands               2371
Norway                    1086
Poland                     341
Portugal                  1519
RSA                         58
Saudi Arabia                10
Singapore                  229
Spain                     2533


United Kingdom、UK、U.K.

USA、United States

### 评估重复数据

无需评估重复值。

### 评估无效或错误数据

In [20]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

产品单价和单价理论上不应该存在负数。

In [23]:
original_data[(original_data["Quantity"] <= 0)]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

C开头交易被取消，是这样的都显示负数吗？

In [26]:
original_data[(original_data["Quantity"] <= 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

不以C开头的数量为负数的交易，unitprice都是0吗？

In [29]:
original_data[
(original_data["Quantity"] <= 0) 
& (original_data["InvoiceNo"].str[0] != "C") 
&(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

由此可见，`Quantity`为负数的情况下，要么`InvoiceNo`以C开头表示已取消，要么`UnitPrice`为0，不贡献销售，不在后续分析范围内，可以删掉。

In [31]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

坏账调整，不是实际销售数据。

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括:
- `InvoiceDate`变量的数据类型转换为为日期时间
- `CustomerID`变量的数据类型转换为字符串
- `Description`变量缺失的观察值删除
- `Country`变量值`"USA"`替换为`"United states"`
- `Country`变量值`"U.K."`、`"UK"`替换为`"United Kingdom"` 
- `Quantity`变量值为负数的观察值删除
- `UnitPrice`变量值为负数的观察值删除

为了区分开经过清理的数据和原始的数据，我们创建新的变量`cleaned_data`，让它为`original_data`复制出的副本。
我们之后的清理步骤都将被运用在`cleaned data`上。

In [46]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [47]:
# InvoiceDate变量的数据类型转换为为日期时间
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

In [48]:
# CustomerID变量的数据类型转换为字符串
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [49]:
# 去掉CustomerID里的.0
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,-2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

In [52]:
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[540455 rows x 8 columns]

In [53]:
# Description变量缺失的观察值删除
cleaned_data.dropna(subset=["Description"], inplace=True)
cleaned_data["Description"].isnull().sum()

np.int64(0)

In [54]:
# Country变量值"USA"替换为"United states"
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United states"})

In [56]:
cleaned_data[cleaned_data["Country"] == "USA"]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [57]:
# Country变量值"U.K."、"UK"替换为"United Kingdom"
cleaned_data["Country"] = cleaned_data["Country"].replace({"U.K.":"United Kingdom"})
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK":"United Kingdom"})
len(cleaned_data[(cleaned_data["Country"] == "U.K.") | (cleaned_data["Country"] == "UK")])

0

`Quantity`变量值为负数的观察值删除，并检查替换后`Quantity`变量为负数的个数。

In [58]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

`UnitPrice`变量值为负数的观察值删除，并检查替换后`UnitPrice`变量为负数的个数。。

In [59]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

In [73]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_412\1440745655.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom

In [74]:
import os

# 使用相对路径
file_path = './e_commerce_cleaned.csv'

# 检查文件是否存在
print(os.path.exists(file_path))

True


In [75]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_412\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom